### PR TITLE
Faster side effects

### PR DESCRIPTION
**Pull request by ppedrot**
This branch fixes more slowness issues observed in [bug #5382](https://coq.inria.fr/bugs/show_bug.cgi?id=5382), that arise when generating a lot of internal lemmas in a proof through calls to abstract. The issues are the following:

- The most serious one is the fact that most of the time, the kernel hashconses twice opaque terms to be stored in the opaque tab. This should probably be backported to 8.6.
- Two minor issues about algorithmic, in one case the costly recomputation of the size of a map when we can store it, and in another repeatedly crawling a huge list for nothing.

@gares should probably be interested by these patches as they gravitate around the side-effect machinery.

### CODE REVIEW

**Comment by gares** [comment]
The commit message is not informative: *why* doing hcons there is granting it is done only once?
I think I get it now, and I agree it is the right thing to do, but it should be said in the commit message.

Also, there is commented code in your last commit that serves no purpose.

File: kernel/opaqueproof.ml [hunk 1]
```diff
@@ -42,21 +52,25 @@ let set_indirect_univ_accessor f = (get_univ := f)
...
-      let cu = Future.chain ~pure:true cu (fun (c, u) -> hcons_constr c, u) in
-      let id = Int.Map.cardinal prfs in
-      let prfs = Int.Map.add id (d,cu) prfs in
-      let ndp =
-        if DirPath.equal dp odp then odp
-        else if DirPath.equal odp DirPath.initial then dp
+      (** Uncomment to check dynamically that all terms turned into
+          indirections are hashconsed. *)
+(* let check_hcons c = let c' = hcons_constr c in assert (c' == c); c in *)
+(* let cu = Future.chain ~pure:true cu (fun (c, u) -> check_hcons c; c, u) in *)
```
Comment (gares) [inline kernel/opaqueproof.ml]: Why is it executed twice?

**Review by gares — CHANGES_REQUESTED** [review]
The commit message his not explaining why the new code is better (and is correct), nor why the old code is doing something twice.

**Comment by ppedrot** [comment]
> The commit message is not informative: why doing hcons there is granting it is done only once?
I think I get it now, and I agree it is the right thing to do, but it should be said in the commit message.

What would you write? All codepath already hashcons the considered term, except for section closing.

> IMHO: it is a total waste of time to backport an optimization. In 2 months we freeze. Will there be a point release of 8.6 by then ?

I'm not really pushing for  the backport, but I was under the impression that the current behaviour was a bug, and we're still fixing bugs in 8.6, aren't we?

**Comment by Zimmi48** [comment]
FYI @maximedenes kind of announced a patch-level release quite soon in https://github.com/coq/coq/pull/220#issuecomment-288741675

**Comment by gares** [comment]
I had the impression the inefficiency is:
- at abstract time the proof term gets hashconsed (if there is no section)
- then such proof term is inlined and hash consed again when the main term is hashconsed

If I'm wrong please explain the code path that resulted into 2 hashconsing and why moving the call to hashconsing at discharge time solves it.  Also, how do you grant all terms are hashconsed (i.e. if there is no section)?

**Comment by ppedrot** [comment]
@gares: There is indeed such an inefficiency, but it is orthogonal to the one solved by this patch.

The problem solved by the current patch is actually very simple. The `Opaqueproof.turn_indirect` function proactively hashconses a term before putting it into the table. It is called only and exactly on opaque constants when outside a section. Problem is, most of the code paths building up the constant already hashcons the term on their side, as witnessed e.g. in `Term_typing.infer_declaration`. So that in practice, when outside a section, declaring an opaque constant triggers a double hashconsing of the proof term. As our hashconsing algorithm is dumb, i.e. O(n), this turns out to be costly (N.B: hashconsing of an already hashconsed term should be O(1) in an ideal world, but **Marshal**). The only path that did not so was `Cooking` so that it needed to be fixed accordingly.

TL;DR: this patch moves the responsibility of hashconsing opaque definitions from `Opaqueproof` to its clients.

**Comment by gares** [comment]
> TL;DR: this patch moves the responsibility of hashconsing opaque definitions from Opaqueproof to its clients.

This looks like a very fragile API.  Why can't we just say that the only component doing hash consing is Opaqueproof (that is at the very bottom of the stack)?  And remove all other calls to hashcons?

**Comment by ppedrot** [comment]
The problem is that not all definition go through Opaqueproof. Furthermore, there is some inversion of control lacking to permit such a thing easily. Thus the current patch is the more straightforward solution I can think of. We are in the kernel, thus we can allow to rely on careful code review and possibly dynamic checking as commented in the source.

**Comment by tchajed** [comment]
This PR doesn't include a test case. I think it would be good for someone to follow Paul's suggestion and reduce the example to the Fano plane; the original reporter is in the best position to do so but somebody else could probably figure it out.

**Comment by ppedrot** [comment]
@tchajed In the meanwhile, could we merge this PR (even ideally in 8.6)? I have to maintain a separate branch to check slowness of particular files.

@gares I don't really understand what you're complaining about. The commit message does explain the issue, and the patch adds a comment to the `Opaqueproof` API.

**Comment by tchajed** [comment]
@ppedrot: Merging seems fine, but thanks for asking! I'll come back to this at some point and see if I can write a test case.

**Review by gares — APPROVED** [review]

File: kernel/term_typing.ml [hunk 2]
```diff
@@ -186,7 +192,8 @@ let rec unzip ctx j =
       unzip ctx { j with uj_val = mkApp (mkLambda (n,ty,j.uj_val),arg) }
 
 let hcons_j j =
-  { uj_val = hcons_constr j.uj_val; uj_type = hcons_constr j.uj_type} 
+  (** Do not hashcons type: it is not used by the callers of this function. *)
```
Comment (maximedenes) [inline kernel/term_typing.ml]: Why is it returned, then?

File: kernel/term_typing.ml [hunk 3]
```diff
@@ -507,7 +514,9 @@ let translate_local_assum env t =
     t
 
 let translate_recipe env kn r =
-  build_constant_declaration kn env (Cooking.cook_constant env r)
+  let (_, dir, _) = Constant.repr3 kn in
+  let hcons = DirPath.is_empty dir in
+  build_constant_declaration kn env (Cooking.cook_constant ~hcons env r)
```
Comment (maximedenes) [inline kernel/term_typing.ml]: Please document that we're hash consing only outside of sections, and why. Imagine an engineer reads this code.

File: kernel/term_typing.ml (hunk 2 above)
Comment (ppedrot) [inline kernel/term_typing.ml]: It seems that at the beginning, it was passed to some function but that code disappeared in between. I'm going to remove it I think.

Comment (ppedrot) [inline kernel/term_typing.ml]: Ah, no, it is still passed to a function but the result is not used. Not sure what to do.

File: kernel/term_typing.ml (hunk 3 above)
Comment (ppedrot) [inline kernel/term_typing.ml]: Added.

File: kernel/term_typing.ml (hunk 2 above)
Comment (maximedenes) [inline kernel/term_typing.ml]: Why do you say it is not used? It is used by `check_cast` in `judge_of_cast`, isn't it? I didn't look further.

Comment (ppedrot) [inline kernel/term_typing.ml]: Yes, but it is irrelevant for this function that the term is hashconsed. The type is not stored anywhere afterwards.

Comment (maximedenes) [inline kernel/term_typing.ml]: How about renaming `hcons_j` to `hcons_term_in_judgement` or something similar? The name is a bit long but it reflects the non-standard behavior (hash consing only the term).

Comment (ppedrot) [inline kernel/term_typing.ml]: It's only used once. I think I'll rather inline it.

Comment (maximedenes) [inline kernel/term_typing.ml]: Ok, sounds good too.
